### ISSUE TITLE
Heretic: desync in h5f3-803

### ISSUE THREAD

**Issue by chungy** [chocolate-doom/chocolate-doom#827]
As reported on [Doomworld](https://www.doomworld.com/vb/post/1693502), this demo exhibits a major desync bug, in recent versions of the master branch the player ceases to make any reasonable actions.

I've bisected the engine and it seems the problematic commit is eccc7950039820344a10a3a330ea39637c301370 -- apparently not as robust as it was assumed.

**Comment by CapnClever** [comment]
This is a duplicate of #299.

**Comment by chungy** [comment]
Hmm, same demo, but different issues related to it.

**Comment by haleyjd** [comment]
It is probable that the argument passing of multiple P_Random calls is to blame, as only Raven's code does that. This will require comparison with the disassembly.

**Comment by haleyjd** [comment]
Hexen should be equally affected unless the argument order emission changed in Watcom between the compiles, so closing this might be premature.

**Comment by jmtd** [comment]
This is amazing, imho. My understanding is that the argument order is undefined in C, so it is compiler-specific choice, and I thought I'd tested the ordering we went for (L-R), but I couldn't have tested it hard enough. Anyway, looks like R-L was the choice Watcom made as well as GCC and LLVM. I wonder if some early compiler (perhaps Watcom) made a choice and that has become a de-facto standard amongst compiler implementors, or if there's some logic behind R-L that I haven't fathomed out. Anyway, excellent work by @nukeykt !

> Hexen should be equally affected unless the argument order emission changed in Watcom between the compiles, so closing this might be premature.

We only tagged this bug Heretic, so we either reopen this and tag it Hexen, or clone it. I'll do the latter, for some reason I prefer having the satisfaction of this one closed. But thank you for pointing out that Hexen is likely still affected.
